### PR TITLE
encrypt-data: Don't recommend AES-CBC

### PR DESCRIPTION
**Pull request by hoskeri**
CBC is not recommended any more due to vulnerability
to padding oracle attacks.

Promote secretbox instead.

### CODE REVIEW

File: content/en/docs/tasks/administer-cluster/encrypt-data.md [hunk 1]
```diff
@@ -78,9 +77,9 @@ read that resource will fail until it is deleted or a valid decryption key is pr
 Name | Encryption | Strength | Speed | Key Length | Other Considerations
 -----|------------|----------|-------|------------|---------------------
 `identity` | None | N/A | N/A | N/A | Resources written as-is without encryption. When set as the first provider, the resource will be decrypted as new values are written.
-`aescbc` | AES-CBC with PKCS#7 padding | Strongest | Fast | 32-byte | The recommended choice for encryption at rest but may be slightly slower than `secretbox`.
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/encrypt-data.md:81]: Why the reordering? I don't think it's strongest-first - `identity` is not an effective confidentiality mechanism!

**Review by sftim — COMMENTED** [review]
This change would seem to actually recommend KMS where available. I'm good with that FWIW.

File: content/en/docs/tasks/administer-cluster/encrypt-data.md (hunk 1 above)
Comment (celestehorgan) [reply content/en/docs/tasks/administer-cluster/encrypt-data.md:81]: +1, why the reordering? The strongest seems to be the last in the table, `kms`, which.....

Comment (hoskeri) [reply content/en/docs/tasks/administer-cluster/encrypt-data.md:81]: That's a good point. Do you think its better separated into two tables? "Recommended" which contains (secretbox, AES-GCM, and KMS), and "Others" which contains (CBC, identity)?

Comment (sftim) [reply content/en/docs/tasks/administer-cluster/encrypt-data.md:81]: I would change “Strongest” to “Weak”, update the Other Considerations column as well, and not change the order. That'd be all that I think we need to do.

**Comment by zq-david-wang** [comment]
The doc also mention that KMS mode is using "AES-CBC with PKCS#7 padding", is this safe/correct?

**Comment by zq-david-wang** [comment]
https://github.com/kubernetes/kubernetes/blob/ee5df7cbcfffaddece66944a528fccb07430d29a/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go#L115-L131
According those lines,  the codes seems to "Decrypt without allowing the decryptor to verify or remove padding" which should be able to reduce the threat. Am I right?

File: content/en/docs/tasks/administer-cluster/encrypt-data.md (hunk 1 above)
Comment (zq-david-wang) [reply content/en/docs/tasks/administer-cluster/encrypt-data.md:81]: kms also use aes-cbc mode, only difference is that the encryption key is managed by external KMS service. And this padding attack can carry out "without knowing the key"

File: content/en/docs/tasks/administer-cluster/encrypt-data.md [hunk 2]
```diff
@@ -78,9 +77,9 @@ read that resource will fail until it is deleted or a valid decryption key is pr
...
 `secretbox` | XSalsa20 and Poly1305 | Strong | Faster | 32-byte | A newer standard and may not be considered acceptable in environments that require high levels of review.
 `aesgcm` | AES-GCM with random nonce | Must be rotated every 200k writes | Fastest | 16, 24, or 32-byte | Is not recommended for use except when an automated key rotation scheme is implemented.
```
Comment (graingert) [inline content/en/docs/tasks/administer-cluster/encrypt-data.md:81]: What does this 200k writes refer to? Doesn't kubenetes use a fresh IV for each document?

**Comment by graingert** [comment]
what about replacing aescbc with aesgcmsiv? https://datatracker.ietf.org/doc/html/rfc8452

**Comment by zq-david-wang** [comment]
I think the official doc is missing threat model definition.  After clarify which thread models are under consideration and which models are explicitly excluded, maybe aes-cbc is still the "strongest"

**Comment by sftim** [comment]
> what about replacing aescbc with aesgcmsiv? https://datatracker.ietf.org/doc/html/rfc8452

@graingert that'd be a feature request against [kubernetes](https://github.com/kubernetes/kubernetes). In this repo, we mostly document the current, observable behavior of released code.

**Comment by kbhawkey** [comment]
Page preview:
https://deploy-preview-29102--kubernetes-io-main-staging.netlify.app/docs/tasks/administer-cluster/encrypt-data/#providers

👋 @hoskeri .
Just checking, is there more work to do on this pull request?

**Comment by zq-david-wang** [comment]
aescbc (AES-CBC with PKCS#7 padding) is weak ==> kms is weak

**Review by sftim — COMMENTED** [review]
Let's make this change. Further tidying is welcome.
/lgtm
/approve

**Pull request merged**